### PR TITLE
Improved zh-CN translation for Haskell

### PR DESCRIPTION
**Pull request by chadluo**
Improved translation. Added snippets from original version.

### CODE REVIEW

File: zh-cn/haskell-cn.html.markdown [hunk 1]
```diff
@@ -5,24 +5,24 @@ contributors:
...
 
-Haskell 被设计成一种实用的纯函数式编程语言。它因为 monads 及其类型系统而出名，但是我回归到它本身因为。Haskell 使得编程对于我而言是一种真正的快乐。
+Haskell 是一门实用的函数式编程语言，因其 Monads 与类型系统而闻名。而我使用它则是因为它异常优雅。用 Haskell 编程令我感到非常快乐。
 
 ```haskell
--- 单行注释以两个破折号开头
-{- 多行注释像这样
-   被一个闭合的块包围
+-- 单行注释以两个减号开头
+{- 多行注释像这样被一个闭合的块包围
```
Comment (geoffliu) [inline zh-cn/haskell-cn.html.markdown]: 我比较喜欢原来的写法，把这个多行注释的例子。。。分成多行

File: zh-cn/haskell-cn.html.markdown [hunk 2]
```diff
@@ -45,73 +45,80 @@ not False -- True
...
 "Hello " ++ "world!" -- "Hello world!"
 
 -- 一个字符串是一系列字符
+['H', 'e', 'l', 'l', 'o'] -- "Hello"
 "This is a string" !! 0 -- 'T'
 
 
 ----------------------------------------------------
--- 列表和元组
+-- 数组和元组
```
Comment (geoffliu) [inline zh-cn/haskell-cn.html.markdown]: 这里应该用列表。Haskell用的是单链表。

File: zh-cn/haskell-cn.html.markdown [hunk 3]
```diff
@@ -262,36 +267,32 @@ for [0..5] show
...
 say Blue = "You are Blue!"
 say Green =  "You are Green!"
 
 -- 你的数据类型也可以有参数：
-
 data Maybe a = Nothing | Just a
 
--- 类型 Maybe 的所有
+-- 这些都是 Maybe 类型：
 Just "hello"    -- of type `Maybe String`
```
Comment (geoffliu) [inline zh-cn/haskell-cn.html.markdown]: "of type" 也翻译一下？

File: zh-cn/haskell-cn.html.markdown [hunk 4]
```diff
@@ -300,58 +301,54 @@ Nothing         -- of type `Maybe a` for any `a`
...
+-- 输入一些文本，对其调用一个函数，并打印输出。
 
 countLines :: String -> String
 countLines = show . length . lines
 
 main' = interact countLines
 
--- 你可以考虑一个 `IO()` 类型的值，当做一系列计算机所完成的动作的代表，
--- 就像一个以命令式语言编写的计算机程序。我们可以使用 `do` 符号来把动作链接到一起。
+-- 你可以认为一个 `IO ()` 类型的值是表示计算机做的一系列操作，类似命令式语言。
```
Comment (geoffliu) [inline zh-cn/haskell-cn.html.markdown]: 这里用“以为”有点怪，“假设”会不会自然一点？

File: zh-cn/haskell-cn.html.markdown [hunk 5]
```diff
@@ -300,58 +301,54 @@ Nothing         -- of type `Maybe a` for any `a`
...
 --   main = sayHello
 
--- 让我们来更好地理解刚才所使用的函数 `getLine` 是怎样工作的。它的类型是：
+-- 让我们来更进一步理解刚才所使用的函数 `getLine` 是怎样工作的。它的类型是：
 --    getLine :: IO String
--- 你可以考虑一个 `IO a` 类型的值，代表一个当被执行的时候
--- 将产生一个 `a` 类型的值的计算机程序（除了它所做的任何事之外）。我们可以保存和重用这个值通过 `<-`。
--- 我们也可以写自己的 `IO String` 类型的动作：
-
+-- 你可以认为一个 `IO a` 类型的值代表了一个运行时会生成一个 `a` 类型值的程序（可能还有其它行为）。
```
Comment (geoffliu) [inline zh-cn/haskell-cn.html.markdown]: 还有这里的“认为”

**Comment by geoffliu** [comment]
Thanks, this is some great work! Only a few minor comments, mostly the "array" vs "list" comment, then we can merge this.

File: zh-cn/haskell-cn.html.markdown (hunk 2 above)
Comment (chadluo) [inline zh-cn/haskell-cn.html.markdown]: 对，这里应当是列表。翻译时有点想当然了。

File: zh-cn/haskell-cn.html.markdown (hunk 4 above)
Comment (chadluo) [inline zh-cn/haskell-cn.html.markdown]: 这里是「认为」，您可能看错了。

**Comment by chadluo** [comment]
Thank you @geoffliu for your supervision. I have mixed up with lists and arrays and I am fixing that, and now working on a fix of it.

File: zh-cn/haskell-cn.html.markdown (hunk 4 above)
Comment (geoffliu) [inline zh-cn/haskell-cn.html.markdown]: 哦，是我打错了。我是说“认为”不是很合适。当你说“我认为”的时候，是表示有考虑过，是你判断的结果。但是这里你想说的意思是假设，把 `IO ()` 当作什么，之前并没有深思熟虑。所以我想说把认为改成“把。。。当成”或者“把。。。想成”

Comment (geoffliu) [inline zh-cn/haskell-cn.html.markdown]: @chadluo 你觉得这样改合不合适？

**Comment by geoffliu** [comment]
Hey @chadluo, I'm gonna merge this PR for now, since it's already a huge improvement from previous version. I still don't agree with your use of "认为" but that's a minor issue we can talk about later. :)

**Pull request merged**